File: A1012/src/Main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.StringTokenizer;

/**
 * A1012: The Best Rank
 * <https://pintia.cn/problem-sets/994805342720868352/problems/994805502658068480>
 * 
 * @author johnniang
 *
 */
public class Main {

	public static void main(String[] args) {

		FastReader reader = new FastReader();
		int N = reader.nextInt();
		int M = reader.nextInt();

		List<Student> students = new ArrayList<>();

		for (int i = 0; i < N; i++) {
			Student student = new Student();
			student.number = reader.next();
			student.c = reader.nextInt();
			student.m = reader.nextInt();
			student.e = reader.nextInt();
			student.a = (student.c + student.m + student.e) / 3;
			students.add(student);
		}

		// Sort
		students.sort(cSort());
		Map<String, Integer> cRankMap = getRankMap(students);

		students.sort(mSort());
		Map<String, Integer> mRankMap = getRankMap(students);

		students.sort(eSort());
		Map<String, Integer> eRankMap = getRankMap(students);

		students.sort(aSort());
		Map<String, Integer> aRankMap = getRankMap(students);

		for (int i = 0; i < M; i++) {
			String number = reader.next();
			Integer cRank = cRankMap.get(number);
			if (cRank == null) {
				System.out.println("N/A");
				break;
			}

			int mRank = mRankMap.get(number);
			int eRank = eRankMap.get(number);
			int aRank = aRankMap.get(number);

			if (eRank < mRank) {
				System.out.println((eRank + 1) + " E");
			} else if (mRank < cRank) {
				System.out.println((mRank + 1) + " M");
			} else if (cRank < aRank) {
				System.out.println((cRank + 1) + " C");
			} else {
				System.out.println((aRank + 1) + " A");
			}
		}

	}

	static class FastReader {
		BufferedReader br;
		StringTokenizer st;

		public FastReader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}

		String next() {
			while (st == null || !st.hasMoreElements()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			return st.nextToken();
		}

		int nextInt() {
			return Integer.parseInt(next());
		}

		long nextLong() {
			return Long.parseLong(next());
		}

		double nextDouble() {
			return Double.parseDouble(next());
		}

		String nextLine() {
			String str = "";
			try {
				str = br.readLine();
			} catch (IOException e) {
				e.printStackTrace();
			}
			return str;
		}
	}

	static Map<String, Integer> getRankMap(List<Student> students) {
		if (students == null || students.isEmpty()) {
			return Collections.emptyMap();
		}
		Map<String, Integer> rankMap = new HashMap<>();
		for (int i = 0; i < students.size(); i++) {
			rankMap.put(students.get(i).number, i);
		}
		return rankMap;
	}

	static void print(List<Student> students) {
		students.forEach(student -> {
			System.out.println(student.number + " " + student.c + " " + student.m + " " + student.e + " " + student.a);
		});
	}

	static Comparator<Student> cSort() {
		return (left, right) -> {
			return right.c - left.c;
		};
	}

	static Comparator<Student> mSort() {
		return (left, right) -> {
			return right.m - left.m;
		};
	}

	static Comparator<Student> eSort() {
		return (left, right) -> {
			return right.e - left.e;
		};
	}

	static Comparator<Student> aSort() {
		return (left, right) -> {
			return right.a - left.a;
		};
	}

	static class Student {
		String number;
		int c;
		int m;
		int e;
		int a;
	}
}

File: A1012/main.cpp
#include <iostream>
#include <algorithm>
#include <math.h>

using namespace std;

// A1012:The Best Rank
// <https://pintia.cn/problem-sets/994805342720868352/problems/994805502658068480>

struct Student {
    int id;
    int a;
    int c;
    int m;
    int e;
};

bool aCompare(Student& left, Student& right) {
    return left.a > right.a;
}

bool cCompare(Student& left, Student& right) {
    return left.c > right.c;
}

bool mCompare(Student& left, Student& right) {
    return left.m > right.m;
}

bool eCompare(Student& left, Student& right) {
    return left.e > right.e;
}

void print(Student* student, int length) {
//    for(int i = 0; i < length; i ++) {
//        cout << student->id << " " << student->c << " " << student->m << " " << student->e << " " << student->a << endl;
//        student++;
//    }
//    cout << endl;
}

const char courseMap[4] = {'A', 'C', 'M', 'E'};

int aRank[1000000] = {0};
int cRank[1000000] = {0};
int mRank[1000000] = {0};
int eRank[1000000] = {0};

int main() {
    int N;
    int M;
    cin >> N >> M;

    Student students[2001];

    for(int i = 0; i < N; i++) {
        cin >> students[i].id >> students[i].c >> students[i].m >> students[i].e;
        students[i].a = (students[i].c + students[i].m + students[i].e) / 3;
    }

    // Sort by a;
    sort(students, students + N, aCompare);

    print(students, N);

    for(int i = 0; i < N; i++) {
        // Handle the same rank
        if(i != 0 && students[i - 1].a == students[i].a) {
            aRank[students[i].id] = aRank[students[i - 1].id];
            continue;
        }
        aRank[students[i].id] = i + 1;
    }

    // Sort by c;
    sort(students, students + N, cCompare);

    print(students, N);

    for(int i = 0; i < N; i++) {
        // Handle the same rank
        if(i != 0 && students[i - 1].c == students[i].c) {
            cRank[students[i].id] = cRank[students[i - 1].id];
            continue;
        }
        cRank[students[i].id] = i + 1;
    }

    sort(students, students + N, mCompare);

    print(students, N);

    for(int i = 0; i < N; i++) {
        // Handle the same rank
        if(i != 0 && students[i - 1].m == students[i].m) {
            mRank[students[i].id] = mRank[students[i - 1].id];
            continue;
        }
        mRank[students[i].id] = i + 1;
    }

    sort(students, students + N, eCompare);

    print(students, N);

    for(int i = 0; i < N; i++) {
        // Handle the same rank
        if(i != 0 && students[i - 1].e == students[i].e) {
            eRank[students[i].id] = eRank[students[i - 1].id];
            continue;
        }
        eRank[students[i].id] = i + 1;
    }

    for(int i = 0; i < M; i++) {
        int id = 0;
        cin >> id;

        int a = aRank[id];
        int c = cRank[id];
        int m = mRank[id];
        int e = eRank[id];

        if(a == 0.0 || c == 0 || m == 0 || e == 0) {
            cout << "N/A" << endl;
            continue;
        }

        int k = 0;
        int min = a;
        if(c < min) {
            min = c;
            k = 1;
        }

        if(m < min) {
            min = m;
            k = 2;
        }

        if(e < min) {
            min = e;
            k = 3;
        }

        cout << min << " " << courseMap[k] << endl;
    }

    return 0;
}
